torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+            raise UsageError(f"At least one of the inputs does not have an attribute '{attr}'.") from error
+
+
+def _compare_tensors(
+    tensor1: torch.Tensor,
+    tensor2: torch.Tensor,
+    asserter: Callable[[torch.Tensor, torch.Tensor], None],
+    **strict_attributes: bool,
+) -> None:
+    if any(tensor.dtype in (torch.complex32, torch.complex64, torch.complex128) for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for complex tensors is not supported yet.")
+    if any(tensor.is_quantized for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for quantized tensors is not supported yet.")
+    if any(tensor.is_sparse for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for sparse tensors is not supported yet.")
+
+    _compare_tensors_meta(tensor1, tensor2, **strict_attributes)
+
+    if tensor1.device != tensor2.device:
+        tensor1 = tensor1.cpu()
+        tensor2 = tensor2.cpu()
+
+    if tensor1.dtype != tensor2.dtype:
+        dtype = _get_comparison_dtype(tensor1, tensor2)
+        tensor1 = tensor1.to(dtype)
+        tensor2 = tensor2.to(dtype)
+
+    asserter(tensor1, tensor2)
+
+
+def _tensors_equal_asserter(tensor1: torch.Tensor, tensor2: torch.Tensor):
+    if not torch.all(torch.eq(tensor1, tensor2)):
+        raise AssertionError("ADDME")
+
+
+@_hide_internal_traceback
+def assert_tensors_equal(
+    tensor1: torch.Tensor, tensor2: torch.Tensor, *, warn_floating_point: bool = True, **strict_attributes: bool
+) -> None:
+    """Assert that two tensors are equal.
+
+    Before the values are compared, the ``shape`` and the ``dtype`` of the tensors are checked for equality. You can
+    change this behavior with the ``strict_attributes`` parameter.
+
+    Depending on the strictness some additional steps may be performed before the value comparison:
+
+    - If the tensors don't live in the same device memory they are moved to CPU memory.
+    - If the tensors don't have the same ``dtype``, they are converted to the one determined by
+        :func:`torch.promote_types`.
+
+    Args:
+        tensor1 (torch.Tensor): First tensor.
+        tensor2 (torch.Tensor): Second tensor.
+        warn_floating_point (bool): Emit a warning if at least one of the inputs is a floating point tensor.
+        **strict_attributes (bool): Set strictness for attribute equality checking. If omitted, this defaults to
+            ``shape=True, dtype=True``.
+
+    Raises:
+        AssertionError: If the specified attributes do not match.
+        AssertionError: If the tensors are not equal.
+    """
+    is_floating_point = tensor1.dtype.is_floating_point or tensor2.dtype.is_floating_point
+    if is_floating_point and warn_floating_point:
+        warnings.warn(
+            "Due to the limitations of floating point arithmetic, comparing floating-point tensors for equality is not "
+            "recommended. Use the respective 'allclose' variant instead.",
```

Comment:
In this case we can name and link assert_tensors_allclose directly instead of alluding to it.

Reply:
Problem is this warning will also be emitted if its is triggered by `assert_equal` later. In that case we should link to `assert_allclose` instead of `assert_tensors_allclose`. I though about an internal `_name` parameter passed from teh top message to avoid this, but wasn't sure if this is a good approach. Thoughts?

Reply:
I would prefer to replicate the lines in each function to emit an appropriate message.